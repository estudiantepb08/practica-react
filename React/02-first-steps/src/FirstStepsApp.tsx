import { MyAwesomeApp } from "./MyAwesomeApp";
import { ItemCounter } from "./shopping-cart/ItemCounter";

interface ItemInCar {
  productName: string;
  quantity: number;
}

const itemInCar: ItemInCar[] = [
  { productName: 'Nintendo Switch 2', quantity: 1 },
  { productName: 'Pro Controller', quantity: 2 },
  { productName: 'Super Smash', quantity: 3 },
]

export function FirstStepsApp() {
  return (
    <>
      <h2>Hello World!</h2>
      <p>My first React app.</p>
      <button>Click me</button>
      <div>
        <h2>Hola dentro de un div </h2>
      </div>
      <hr />
      <h1>Carrito de Compras</h1>
      {
        itemInCar.map(({ productName, quantity }) => (
          <ItemCounter key={productName} name={productName} quantity={quantity} />
        ))
      }
      <hr />
      <MyAwesomeApp />
    </>
  )
}