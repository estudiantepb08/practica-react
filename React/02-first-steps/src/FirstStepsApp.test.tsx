import { describe, test, expect, vi, afterAll } from 'vitest';
import { render, screen } from '@testing-library/react';
import { FirstStepsApp } from './FirstStepsApp';

const mockItemCounter = vi.fn((props: unknown) => {
    return (<div data-testid="ItemCounter" />);
});

vi.mock('./shopping-cart/ItemCounter', () => ({
    ItemCounter: (props: unknown) => mockItemCounter(props),
}));

/* Otra forma de hacer el mock
vi.mock('./shopping-cart/ItemCounter', () => {
    return {
        ItemCounter: (props: unknown) => <div data-testid="ItemCounter" name={props.name} quantity={props.quantity} />//Mocked ItemCounter</div>
    };
});
*/

describe('FirstStepsApp', () => {

    /** Limpia todos los mock que se ha ejecutado en las pruebas */
    afterAll(() => {
        vi.clearAllMocks();
    });

    test('should match snapshot', () => {
        const container = render(<FirstStepsApp />);
        expect(container).toMatchSnapshot();

    });

    test('should render the correct number of ItemCounter Components', () => {
        render(<FirstStepsApp />);
        const itemCounter = screen.getAllByTestId('ItemCounter');
        console.log(itemCounter.length);
        expect(itemCounter.length).toBe(3);
        screen.debug();
    });

    test('should render ItemCounter with the correct props', () => {
        render(<FirstStepsApp />);
        expect(mockItemCounter).toHaveBeenCalledTimes(9);
        expect(mockItemCounter).toHaveBeenCalledWith({ name: 'Nintendo Switch 2', quantity: 1 });
        expect(mockItemCounter).toHaveBeenCalledWith({ name: 'Pro Controller', quantity: 2 });
        expect(mockItemCounter).toHaveBeenCalledWith({ name: 'Super Smash', quantity: 3 });
    });
});