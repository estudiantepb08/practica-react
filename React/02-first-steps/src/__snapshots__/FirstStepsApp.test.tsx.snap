// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FirstStepsApp > should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h2>
        Hello World!
      </h2>
      <p>
        My first React app.
      </p>
      <button>
        Click me
      </button>
      <div>
        <h2>
          Hola dentro de un div 
        </h2>
      </div>
      <hr />
      <h1>
        Carrito de Compras
      </h1>
      <div
        data-testid="ItemCounter"
      />
      <div
        data-testid="ItemCounter"
      />
      <div
        data-testid="ItemCounter"
      />
      <hr />
      <h1>
        Demetrio
      </h1>
      <h2>
        Sinisterra
      </h2>
      <p>
        Elden Ring, Smash Bros, Metal Gear
      </p>
      <p>
        6
      </p>
      <p>
        Activo
      </p>
      <h3
        style="background-color: rgb(240, 240, 240); border-radius: 10px; padding: 10px;"
      >
        {"zipCode":"ABC-123","country":"Colombia"}
      </h3>
    </div>
  </body>,
  "container": <div>
    <h2>
      Hello World!
    </h2>
    <p>
      My first React app.
    </p>
    <button>
      Click me
    </button>
    <div>
      <h2>
        Hola dentro de un div 
      </h2>
    </div>
    <hr />
    <h1>
      Carrito de Compras
    </h1>
    <div
      data-testid="ItemCounter"
    />
    <div
      data-testid="ItemCounter"
    />
    <div
      data-testid="ItemCounter"
    />
    <hr />
    <h1>
      Demetrio
    </h1>
    <h2>
      Sinisterra
    </h2>
    <p>
      Elden Ring, Smash Bros, Metal Gear
    </p>
    <p>
      6
    </p>
    <p>
      Activo
    </p>
    <h3
      style="background-color: rgb(240, 240, 240); border-radius: 10px; padding: 10px;"
    >
      {"zipCode":"ABC-123","country":"Colombia"}
    </h3>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
